src/idl_gen_js_ts.cpp

```diff
@@ -661,6 +674,486 @@ class JsTsGenerator : public BaseGenerator {
+              }
+              default: {
+                if (vectortype.enum_def) {
+                  field_type +=
+                      GenPrefixedTypeName(GenTypeName(vectortype, false, true),
+                                          vectortype.enum_def->file);
+                } else {
+                  field_type += vectortypename;
+                }
+                field_type += ")[]";
+                field_val = GenBBAccess() + ".createScalarList(" +
+                            field_binded_method + ", this." + field_name +
+                            "Length())";
+
+                field_pre_offset = Verbose(struct_def) + ".create" +
+                                   MakeCamel(field_name) +
+                                   "Vector(builder, this." + field_name + ")";
+
+                break;
+              }
+            }
+
+            break;
+          }
+
+          case BASE_TYPE_UNION: {
+            field_type +=
+                GetObjApiUnionTypeTS(parser.opts, *(field.value.type.enum_def));
+
+            field_val = GenUnionValTS(field_name, field.value.type);
+            field_pre_offset =
+                "builder.createObjectOffset(this." + field_name + ")";
+            break;
+          }
+
+          default: FLATBUFFERS_ASSERT(0); break;
+        }
+
+        // length 0 vector is simply empty instead of null
+        field_type += is_vector ? "" : "|null";
+      }
+
+      if (!field_pre_offset.empty()) {
+        field_pre_offset =
+            "  const " + field_name + " = " + field_pre_offset + ";";
+      }
+      if (field_post_offset.empty()) { field_post_offset = field_name; }
+
+      ret.push_back(ObjApiData{ field_name, field_val, field_type,
+                                field_pre_offset, field_post_offset,
+                                field_default_val });
+    }
+
+    return ret;
+  }
+
+  // Create a string for each member in data based on fmt.
+  //
+  // For example, with fmt being "public $name: $type", every struct member
+  // will have a string created and added to code_ptr in that format.
+  //
+  // $name for name of the member in camel case
+  //
+  // $type for object api type if it's struct or table, o.t.w it's scalar
+  // types
+  //
+  // $val for the value after unpacking, assuming global this of generated
+  // code is the non object api type $offset offset of the type for packing,
+  // assuming global this is the object being packed
+  //
+  // $default_val for default value of the field
+  //
+  // $pre_offset will be a declaration of a variable with name $name and
+  // will be non-empty and contain the offset for things that can't be
+  // serialized inline such as table
+  //
+  // $post_offset will include the value of variable declared in pre_offset
+  // as well as offset(or values) of things that can be serialized inline
+  // such as offset for struct or value for scalar
+  void FmtObjApi(const std::list<ObjApiData> &data, std::string &code,
```

Comment:
Please generate code much like the code generation that already existed in this file, rather than inventing a new way of doing it based on regex, so that others can maintain this code easily.

Reply:
I changed it into a callback style. While I understand that inconsistency is bad for maintaining, it's hard to follow the exact current style because I would either need to add to the already big `GenStruct` function or repeat a loop multiple times to gather overlapping data. Do you have any suggestions?

Reply:
My suggestion is to write the code much like the rest of the file was already doing, and all the other code generators are doing, meaning not first collecting parameters and formatting them afterwards, but doing it directly where needed, with no additional abstraction. That's most consistent and easiest to change for others. Adding code to an already big function is no concern.